cpp/include/cudf/detail/valid_if.cuh

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/null_mask.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/bit.cuh>
+
+#include <rmm/device_scalar.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+template <size_type block_size, typename InputIterator, typename Predicate>
+__global__ void valid_if_kernel(bitmask_type* output, InputIterator begin,
+                                InputIterator end, Predicate p,
+                                size_type* valid_count) {
+  constexpr size_type leader_lane{0};
+  auto const lane_id{threadIdx.x % warp_size};
+  auto const warp_id{threadIdx.x / warp_size};
+  auto const tid = threadIdx.x + blockIdx.x * gridDim.x;
+  auto i = begin + tid;
+  size_type warp_valid_count{0};
+
+  auto active_mask = __ballot_sync(0xFFFF'FFFF, i < end);
+  while (i < end) {
```

Comment:
Typically with iterators you use equality comparison because it's more generic and flexible (all iterators are equality comparable but not all iterators are relationally comparable).
https://stackoverflow.com/a/6673775/749748

So here this would seem to be more idiomatic.

```suggestion
  while (i != end) {
```

However, it doesn't work, because you are not incrementing by one, you are incrementing by the number of threads. But typically incrementing an iterator beyond the end of the range is undefined behavior...

I get that using iterators buys composability, but I feel like ranges might be more amenable to parallelism while still being composable?  For now, maybe need to document the requirements on the InputIterator type.

Reply:
Yeah, good point. I'll just modify to pass in the `begin` and `size`.